### ISSUE TITLE
Add possibility to train KTO on an unpaired dataset

### ISSUE THREAD

**Issue by maxjeblick** [huggingface/trl#1116]
First of all thanks a lot for the fast addition of KTO loss!

With KTO, you don't need pairwise preferences, see e.g. [here](https://github.com/ContextualAI/HALOs/blob/main/dataloader.py#L404) and the technical report. That is, a dataset consisting of columns `["prompt", "answer", "is_good"]` with binary label `is_good` would be the required input to KTO (the reference implementation still uses pairwise datasets and then flattens it into a list of q&a pairs which is a bit unfortunate imo).

In trl, if I'm not mistaken, you would still need preferred-unpreffered pairs, as the input for the model is constructed with `prompt + chosen` and `prompt + rejected`, respectively,  see [here](https://github.com/huggingface/trl/blob/main/trl/trainer/dpo_trainer.py#L576).

**Comment by alvarobartt** [comment]
Yes, indeed I was also a bit surprised that they chose chosen-rejected pair datasets for showcasing KTO rather than using a dataset with unpaired chosen-rejected rows; I think it was just an unfortunate pick of the datasets 😕 Also I think that porting a separate `DataLoader` for unpaired preference data may do the work? See https://github.com/ContextualAI/HALOs/blob/8d8aac208e20965d8f3212a2633fab86f15861ff/dataloader.py#L402

**Comment by lvwerra** [comment]
Tagging @kashif here :)

**Comment by raghavgarg97** [comment]
do we have any plan to pick this up?would love to give KTO a try

**Comment by maxjeblick** [comment]
Commenting to not mark it as stale.

**Comment by kawine** [comment]
@maxjeblick until this is implemented in TRL, you're welcome to check out the original implementation in https://github.com/ContextualAI/HALOs, which also has stabler training due to how the KL term is estimated

The technical report has some guidance on how to adjust the loss weights if you have a heavy balance of good:bad data or vice-versa.

**Comment by kashif** [comment]
see issue #1181
